File: android/app/src/main/kotlin/com/app/deenly/deenly/assetdownload/AssetManifestFetcher.kt
package com.rnr.deenfocus.assetdownload

import org.json.JSONObject

/**
 * Fetches the tiny top-level catalog and per-pack manifests over HTTPS (ADR-008).
 * Knows nothing about what a "pack" is used for (Tajweed, translations, TTS
 * voices, OCR, ...) — callers decode the per-pack manifest into whatever shape
 * they need.
 */
class AssetManifestFetcher(
    private val transport: AssetTransport = HttpUrlConnectionAssetTransport(),
) {
    fun fetchCatalog(url: String, timeoutMs: Int = 5_000): AssetCatalog {
        val text = transport.fetchJson(url, timeoutMs)
        return try {
            AssetCatalog.parse(JSONObject(text))
        } catch (e: Exception) {
            throw AssetDownloadException.DecodeFailed("catalog.json: ${e.message}")
        }
    }

    /**
     * Fetches a per-pack manifest as a raw [JSONObject] — the consumer (e.g.
     * Tajweed's `TajweedAssetSync`) interprets its own keys (`encoder`,
     * `pronunciationHead`, `sha256`, ...), matching the existing ADR-007 shape.
     */
    fun fetchManifestJson(url: String, timeoutMs: Int = 10_000): JSONObject {
        val text = transport.fetchJson(url, timeoutMs)
        return try {
            JSONObject(text)
        } catch (e: Exception) {
            throw AssetDownloadException.DecodeFailed("manifest is not a JSON object: ${e.message}")
        }
    }
}

File: ios/Runner/AssetDownload/AssetManifestFetcher.swift
import Foundation

/// Fetches the tiny top-level catalog and per-pack manifests over HTTPS (ADR-008).
/// Knows nothing about what a "pack" is used for (Tajweed, translations, TTS voices,
/// OCR, ...) — callers decode the per-pack manifest into whatever shape they need.
final class AssetManifestFetcher {
  private let transport: AssetTransport

  init(transport: AssetTransport = URLSessionAssetTransport()) {
    self.transport = transport
  }

  func fetchCatalog(url: URL, timeout: TimeInterval = 5) throws -> AssetCatalog {
    let data = try transport.fetchJSON(url: url, timeout: timeout)
    do {
      return try JSONDecoder().decode(AssetCatalog.self, from: data)
    } catch {
      throw AssetDownloadError.decodeFailed("catalog.json: \(error.localizedDescription)")
    }
  }

  /// Fetches a per-pack manifest as a raw dictionary — the consumer (e.g. Tajweed's
  /// `TajweedAssetSync`) interprets its own keys (`encoder`, `pronunciationHead`,
  /// `sha256`, ...), matching the existing ADR-007 manifest shape exactly.
  func fetchManifestJSON(url: URL, timeout: TimeInterval = 10) throws -> [String: Any] {
    let data = try transport.fetchJSON(url: url, timeout: timeout)
    guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      throw AssetDownloadError.decodeFailed("manifest is not a JSON object")
    }
    return obj
  }
}
